Add takestring!(x) to create a string from the content of x, emptying it

jakobnissen opened the pull request:
This PR adds the `takestring` function. It currently has two methods:
* The generic `takestring!(x)::String` creates a string from the content of `x`, emptying `x`. It's currently only defined for `Vector{UInt8}`.
* `takestring(::GenericIOBuffer)::String` creates a string from the content of the buffer, and resets the buffer to the initial state.

For the motivation for this function, and discussion, read the comments on this PR further down.

nhz2 commented:
Also, check for uses of `StringMemory` for example:
https://github.com/JuliaLang/julia/blob/f3561ae5af05713480aa9a8166501449305a0339/stdlib/FileWatching/src/pidfile.jl#L283-L288

nhz2 commented on base/strings/string.jl:
```diff
@@ -85,6 +82,36 @@ function String(v::Vector{UInt8})
+
+Create a string, truncating `v` to zero length. If `v` is a `Vector`, further
+modification of `v` will not modify the string.
+When possible, the returned string will reuse the memory of `v`.
```
Normally `Memory` length can't be mutated, so it might not get truncated. This should also 
 say mutating the `Memory` after taking a string might be unsafe.

jakobnissen replied:
This implementation will truncate the memory to size zero, which means it's impossible to mutate the memory.

nhz2 replied:
`jl_genericmemory_to_string` is going to change to not truncate the memory:
https://github.com/JuliaLang/julia/pull/54372#issuecomment-2102885169

jakobnissen commented on base/strings/string.jl:
```diff
@@ -63,12 +63,9 @@ by [`take!`](@ref) on a writable [`IOBuffer`](@ref) and by calls to
-    return ccall(:jl_genericmemory_to_string, Ref{String}, (Any, Int), v, len)
-end
+String(v::AbstractVector{UInt8}) = take_string!(copyto!(StringMemory(length(v)), v))
+String(v::Memory{UInt8}) = take_string!(copy(v))
```
Yes, but `copy(::Memory)` is more efficient.

nhz2 replied:
When testing I get the `copyto!` version being almost twice as fast.

```julia
julia> using BenchmarkTools

julia> function foo!(v::Memory{UInt8})
           len = length(v)
           len == 0 && return ""
           return ccall(:jl_genericmemory_to_string, Ref{String}, (Any, Int), v, len)
       end
foo! (generic function with 1 method)

julia> bar1(v) = foo!(copyto!(Base.StringMemory(length(v)), v))
bar1 (generic function with 1 method)

julia> bar2(v) = foo!(copy(v))
bar2 (generic function with 1 method)

julia> a = Base.StringMemory(1_000_000);

julia> fill!(a, 0x61);

julia> @btime bar1($a);
  34.581 μs (2 allocations: 976.72 KiB)

julia> @btime bar2($a);
  77.967 μs (3 allocations: 1.91 MiB)

```

oscardssmith replied:
apparently our genericmemory_copy is doing something wrong...

nhz2 replied:
I think it is just that `copy` is making a normal memory, not a string-backed memory. Calling `String` on a normal memory does a copy.

jakobnissen replied:
Apologies - I've deleted that method as you're right, it is indeed less efficient.

stevengj commented:
The name makes me think that `take_string!(iobuffer)` should be a replacement/synonym for `String(take!(iobuffer))`?

oscardssmith commented:
I would be in favor of that. Every time I have used an IOBuffer I have had to go to the docs to figure out how to get the data out of it correctly.

jariji commented:
So what's the specification of `String(_)`: does it truncate or not? Or does it depend on the specific argument type? An "it depends" is not a very nice property for generic programming.

jakobnissen commented:
It does not truncate - _except_ when the argument is a `Vector{UInt8}`. Yes, that's bad and in my opinion it shouldn't have been like that. But it's documented, and therefore can't change. I made this PR to make sure more exceptions aren't added.

stevengj commented:
Do we need an underscore here? `takestring!` is pretty readable

oscardssmith commented on base/strings/string.jl:
````diff
@@ -85,6 +82,36 @@ function String(v::Vector{UInt8})
+true
+```
+"""
+function take_string!(v::Memory{UInt8})
````
this should probably get some effect annotations.

oscardssmith commented:
Triage had a long talk about this, https://github.com/JuliaLang/julia/issues/54369, and  https://github.com/JuliaLang/julia/pull/54273. The conclusions we came to are
1. `String(Memory)` should copy
2. `String(Array)` should truncate the array but not the Memory
3. triage was right last time wrt `view` and `reshape` of `Memory`
4. using the an `Array` aliased with another `Array` that is turned into a `String` is UB.
5. `take_string!` is a much better API for IOBuffer et al.

stevengj commented:
The [style guide](https://docs.julialang.org/en/v1/manual/style-guide/) *not* to use underscores when readable, which I think applies here.

topolarity commented:
> 4. using the an Array aliased with another Array that is turned into a String is UB.

Can this be refined to "_mutating_ an Array after its alias has been turned into a String is UB"?

oscardssmith commented:
Yes. observing it is not great but not UB. the only UB is if you mutate it.

nhz2 commented:
`takestring!` can mutate the bytes in the underlying memory of a `Vector` input, it is not just making a `String` view.

jakobnissen commented:
After triage, this PR now contains the following changes:
- [x] Add `takestring!`. This function can be used with `IOBuffer`, `Vector{UInt8}` and `Memory`, returning a `String`. It empties the input given.
- [x] `String(::Memory)` now no longer truncates the memory
- [x] `String(::Vector{UInt8})` now no longer truncates the underlying memory of the string, but instead copies the memory into a string. It still empties the vector, and assigns a new (empty) memory to the vector just like before. Note that this is a change from before, since `String(::Vector{UInt8})` now copies, and therefore is slower. However, see #54434. Another option is to keep the old non-copying behaviour, and accept the larger risk of UB.
- [x] The many different calls to `String(take!(::IOBuffer))` and `String(_unsafe_take!(::IOBuffer))` has been replaced with `takestring!` and `unsafe_takestring!`.
- [ ] Replace _all_ uses of `String(take!(::IOBuffer))`
- [ ] Add tests

stevengj commented on NEWS.md:
```diff
@@ -73,6 +73,7 @@ New library features
 * `tempname` can now take a suffix string to allow the file name to include a suffix and include that suffix in
   the uniquing checking ([#53474])
 * `RegexMatch` objects can now be used to construct `NamedTuple`s and `Dict`s ([#50988])
+* `takestring!(v)` creates a `String` from `v`, truncating `v` and reusing its memory if possible.
```
```suggestion
* `takestring!(v)` creates a `String` from `v` (a `Vector{UInt8}` or `IOBuffer`), truncating `v` and reusing its memory if possible )([#54372]).
```

stevengj commented on base/strings/string.jl:
````diff
@@ -85,6 +80,24 @@ function String(v::Vector{UInt8})
+true
+```
+"""
+takestring!(v::Vector{UInt8}) = String(v)
````
Can we gain anything by having an `unsafe_takestring!(v::Vector{UInt8})` that leaves `v` unchanged, and relies on the caller not to modify it?

jakobnissen replied:
Yes - I've made one locally and is about to change up all the bad uses of takestring!(::Memory) In this PR (which will now fail as I removed that method), but it'll have to wait until Sunday when I get time again :)

jakobnissen commented:
This should be good to go as soon as tests pass. See this comment for a complete list of changes in this PR: https://github.com/JuliaLang/julia/pull/54372#issuecomment-2105059874

KristofferC commented:
The current behavior is already approved by triage (https://github.com/JuliaLang/julia/issues/24388#issuecomment-341546898) so it feels a bit odd that there is a new triage later that just goes and overrides this without any reference (at least from what I can see) to the original decision.

nhz2 commented on base/iobuffer.jl:
````diff
@@ -489,6 +489,62 @@ function take!(io::IOBuffer)
+true
+```
+"""
+function takestring!(io::IOBuffer)
````
This is missing a method for `GenericIOBuffer`.

Also, it is doing an extra copy that can sometimes be avoided.
Maybe the `IOBuffer` changes can be moved to a follow-up PR to make this PR easier to review?

JeffBezanson replied:
Yes, I like having this function but it copies unconditionally, which defeats the purpose. If you're going to copy the data there's no need for any unsafe or `!`.

oscardssmith commented:
@KristofferC It is useful context and was not discussed in the previous triage call. That said, I think none of what triage said in this call contradicts the previous call. Almost all of the decision points for triage here were about the behavior when `Memory` is involved which didn't exist in 2017. The reason re-discussion of such a similar issue was needed is that as of the previous conversation, the solution was to 0 out the length field which was relatively bullet-proof since Julia at the time didn't let you access the underlying fixed length thing upon which `Vector` is built. As of Julia 1.11, `Array` is a "normal" julia object, and `Memory` is the fixed length backing, so the conversation got pushed a layer further down, at which point the previous solution no longer works without some modification.

The one part where the current triage call somewhat diverged from the previous is the decision that a new function that does the combination of removing data from an `IOBuffer` and turning that into a `String` is a good idea. That doesn't seem to have been on the radar as a possible option in the original triage decision, and I do not know why.

topolarity commented:
Is the problem here then that we don't have any way to do a safe runtime "move" out of a Memory?

I'm aware of three options for a copy/performance tradeoff like this:
 - guarantee a copy
 - leave behind an empty object (C++-style "move")
 - leave behind an invalid object (e.g. one which may be potentially freed or whose mutation violates immutability, etc. - triggers UB)

I think our policy on the ownership theft of `String` was that it was allowed because we implemented a runtime "move" in the style of #2. The user might see unexpected mutation/moves, but usage of the left-over object doesn't directly trigger UB.

I'm assuming that's not an option for `Memory`, - because of the assumption that it's always allocated and fixed-size.

nhz2 commented:
`String(Memory)` has to copy according to the docs of `String` since `Memory` is an `AbstractVector{UInt8}` and isn't a `Vector{UInt8}`.

jakobnissen commented:
On @KristofferC 's recommendation, I have now read through the old decisions in #24388, #25241, #25846 and #26093.
I still believe the changes in this PR and #54452 are for the better.

In #24388, the two main points are:
    1. It should not be possible to mutate strings except through `unsafe_`.
    2. Always requiring a copy is a performance disaster

The other PRs are less relevant - they pertain to `codeunits` and `unsafe_wrap`, except #26093 which reiterates that strings should not be mutated, and [also says](https://github.com/JuliaLang/julia/pull/26093#issuecomment-366478405):

> One place where something like String! or string! would be valuable though is to replace String(take!(iobuffer)). Perhaps take!(iobuffer, String). That would allow us to avoid an extra allocation since we know the caller doesn't want the vector, just the string underneath, so the buffer can reuse the vector object (just not its storage).

Which essentially suggests encapsulating this whole sketchy thing with creating strings from mutable data in a single function, as done here.

So, I just don't see how the old triage decisions conflict with the PR here, and given that triage has approved this already, I think we should move forward with it.

To motivate it a little more, I think the status quo is not acceptable because it leads to memory bugs too easily:

Here is a way that works on released versions of Julia unless #54452 is applied to mutate strings
```julia
b = IOBuffer()
write(b, "abc")
v1 = take!(b)
v2 = view(v1, :)
s = String(v1) # "abc"
fill!(v2, 0)
s # now all zeros
```

Here is a bug on master where a vector now is able to read from out of bounds (fixed by #54452):
```julia
v1 = Memory{UInt8}(codeunits("abc"))
v2 = view(v1, 1:3)
s = String(v1)
println(isempty(v2.ref.mem)) # true
v2[1] # accesses v2.ref.mem OOB
```

More broadly speaking, I'm not suggesting that we compromise on our APIs in order to categorically prevent unsafe behaviour. We're not Rust, and need to be practical about what is unsafe. However:

1. This is not some crazy edge case, it's pretty easily accessible memory bugs that can happen by accident way too easily through no obvious fault of the user

2. This is not an API compromise. In my opinion, `takestring!` provided here, and the change in #54452 makes Julia's API _more_ consistent, and less surprising, because it allows users to not have to rely on the weird magic behaviour of truncating `String`, while not providing people with vectors that are sometimes string-backed and sometimes not.

JeffBezanson commented on base/iobuffer.jl:
```diff
@@ -489,6 +489,62 @@ function take!(io::IOBuffer)
 end
 
+"Internal method. Copies the data from the IOBuffer to a string, leaving the
+IOBuffer in an unusable (erroneous) state.
```
How so? It looks like this doesn't modify the iobuffer at all?

jakobnissen replied:
Fixed in the latest commit - it still no longer modifies the buffer, but it now creates a string without copying (and without resetting the buffer to a usable state), which means mutating the buffer is illegal - hence the `unsafe_` name.

JeffBezanson commented:
@jakobnissen Thank you for your tireless efforts here. I do like `takestring!` and I agree with the approach recommended by triage. See comments about ensuring copies are avoided where possible.

From the implementation here the difference between `takestring!` and `unsafe_takestring!` is not clear to me. Can an unsafe version share data in more cases? And my understanding is that the unsafe version would be used internally when we know the IOBuffer has no other references? However in the changes here it is not consistently used that way. Maybe we don't need both functions for IOBuffer?

jakobnissen commented:
Thanks for the feedback. I've now changed both `takestring!` and `unsafe_takestring!` to avoid making copies, where possible. I didn't do that previously, because I mistakenly thought that `IOBuffers`' memory wasn't backed by strings so it would copy anyway. That was weird, because I did know that it was supposed to be zero-copy, but I guess after revising it a few times I lost the forest for the trees.

The `unsafe` variant is a tiny bit faster because a) it assumes the buffer is writable and seekable, and b) it doesn't spend time resetting the buffer to a usable state. In benchmarks, it's only about 10% faster, but still, worth having IMO.

jakobnissen commented:
Bump. What needs to be done to get this merged?

nsajko commented:
FYI FWIW there are unresolved conflicts.

jakobnissen commented:
Bump

stevengj commented on base/iobuffer.jl:
```diff
@@ -489,6 +489,86 @@ function take!(io::IOBuffer)
+"Internal method. This method can be faster than takestring!, because it assumes
+the buffer is writable and seekable, and because it does not reset the buffer
+to a usable state.
+Using the buffer after calling unsafe_takestring! may cause illegal behaviour.
```
```suggestion
Using the buffer after calling unsafe_takestring! may cause undefined behaviour.
```

jakobnissen replied:
Fixed in the latest commit.

stevengj commented on base/iobuffer.jl:
```diff
@@ -489,6 +489,86 @@ function take!(io::IOBuffer)
+    nbytes = filesize(io)
+
+    # If the memory is invalidated (and hence unused), or no bytes, return empty string
+    return if (iszero(nbytes) || io.reinit)
```
Can't this code call `unsafe_takestring!(io)`?

jakobnissen replied:
Re: This comment and the one below: I couldn't remember anymore, and had to re-read the `iobuffer.jl` code. I must admit I find it difficult to understand what invariants the fields of the buffer follow - the precise meaning of the fields is obscure.
Nonetheless, it looks to me like the difference between this computation in `takestring!` and `unsafe_takestring!` is because the latter assumed the input buffer is seekable, which apparently changes whether the beginning of the readable data is given by `.ptr` or `.offset + 1`.
However, it looks like `IOBuffer` is always seekable, since all the public constructors only return instances with `seekable=true`.

So, in the latest commit, I indeed just call `unsafe_takestring!` here, using the same computation.
Let's see if the tests pass.

stevengj commented on base/iobuffer.jl:
```diff
@@ -489,6 +489,86 @@ function take!(io::IOBuffer)
+    return if (iszero(nbytes) || io.reinit)
+        ""
+    else
+        start = io.seekable ? io.offset + 1 : io.ptr
```
Why is this offset computed differently in `unsafe_takestring!`?

jakobnissen replied:
See above.

stevengj commented on base/iobuffer.jl:
```diff
@@ -489,6 +489,75 @@ function take!(io::IOBuffer)
+    # Currently, `IOBuffer`s (as opposed to `GenericIOBuffer`) is always seekable, since all
+    # IOBuffer constructors return seekable buffers, and GenericIOBuffer is internal.
+    # This makes `unsafe_takestring!` valid.
+    s = unsafe_takestring!(io)
```
It seems cleaner and more future-proof for `unsafe_takestring!` to simply use `start = io.seekable ? io.offset + 1 : io.ptr`, like your original implementation of `takestring!` so that it works when `io.seekable == false` — i.e. merge your old code into `unsafe_takestring!` rather than deleting it.

For example, the `take!(io::IOBuffer)` method does not assume this.

jakobnissen replied:
Yeah, that's probably right. However, note that `_unsafe_take!`, which I copied, does assume this.

stevengj commented on base/iobuffer.jl:
```diff
@@ -489,6 +489,75 @@ function take!(io::IOBuffer)
+    end
+    s
+end
+
```
Seems like there should also be a fallback method like:
```jl
takestring!(io::GenericIOBuffer) = String(take!(io))
```

jakobnissen replied:
Good idea. Addressed in last commit.

stevengj commented:
I'm not completely sold on the name `takestring(::Memory)`.   It seems quite different from `takestring!(::IOBuffer)`, since it doesn't "take" anything from the `Memory`.   Maybe `unsafe_wrap(String, ::Memory)` or `unsafe_string(::Memory)`?

jakobnissen commented:
IMO it's fine. It's similar to `unsafe_takestring!`, which doesn't take anything from the IOBuffer either. Or, viewed from another angle, it _does_ take the memory away in the sense that the memory cannot be safely modified after use, because the ownership has been "taken" by someone else.
That being said, I don't mind if it's changed.

nhz2 commented:
After a call to `unsafe_takestring(::Memory)` it is also unsafe to read from the memory because Julia can free the memory. For example, here is an example that leads to a seg fault.

```julia
function foo(n)
    m = Base.StringMemory(n)
    m .= 0x43
    Base.unsafe_takestring(m)
    m
end

m = foo(1_000_000)
GC.gc()
GC.gc()
GC.gc()
GC.gc()
m[1]
```

nhz2 commented on base/strings/annotated.jl:
```diff
@@ -274,7 +274,7 @@ function annotatedstring(xs...)
             print(s, x)
         end
     end
-    str = String(take!(buf))
+    str = unsafe_takestring!(buf)
```
```suggestion
            print(s, x)
        end
    end
    str = takestring!(buf)
```
`buf` can escape this function though `print(s, x)` because `x` can be any type, and `s` contains a reference to `buf`

jakobnissen replied:
While that's true theoretically, I don't think that's an issue realistically. If I understand correctly, this is only a problem if the user overloads `print` to mutate a nonlocal variable such that it stores a reference to the IO object, which is a totally bonkers thing to do - and not at all what `print` is supposed to do.
The issue is also present in many other uses of `unsafe_takestring!` in this PR, but they all seem unlikely to cause issues.

nhz2 replied:
They may be unlikely to cause an issue, but any potential segfault they cause will be very hard to debug or test. I think it would be better to use the safe version unless the code is in some performance critical hot loop.

nhz2 replied:
Someone might try and debug their custom print function using something like https://github.com/JuliaLang/julia/pull/55131 to save local variables to global scope.

jakobnissen commented:
Bump - all comments have been addressed.

nhz2 commented on base/iobuffer.jl:
```diff
@@ -489,6 +489,76 @@ function take!(io::IOBuffer)
+    # we can return an empty string without interacting with the buffer at all.
+    io.reinit && return ""
+
+    s = unsafe_takestring!(io)
```
If the buffer is not writable, this must copy the data into a `StringMemory`, see line 471 in the `take!` function.

jakobnissen replied:
I don't think so, but perhaps I'm wrong. The reason it copies in line 471 is because, if the buffer is not writable, `io.reinit` is not set to `true`, and thus the returned value is an alias of the memory in the buffer. However, this is only safe is one can guarantee the returned `Vector` is never mutated. We can guarantee that when returning `String`, so it's fine that the returned string shares memory with the unwriteable buffer.

nhz2 replied:
This will normally be fine, but the following example seg faults with this PR.

```julia
iow = IOBuffer()
write(iow, zeros(UInt8, 10_000_000))
v1 = take!(iow)
ior = IOBuffer(v1)
s1 = takestring!(ior)
s1 = nothing
GC.gc()
GC.gc()
GC.gc()
v1[1]
```

This is because after a call to `unsafe_takestring(m::Memory{UInt8})` `m` becomes unsafe to read because its memory may get freed by Julia if the string isn't needed.

jakobnissen replied:
Good point, addressed in latest commit

nhz2 commented on base/show.jl:
```diff
@@ -2718,7 +2718,7 @@ function type_depth_limit(str::String, n::Int; maxdepth = nothing)
         prev = di
     end
-    return String(take!(output))
+    return unsafe_takestring!(output)
```
Why is this guaranteed to be safe?

jakobnissen replied:
Because there is no way for this function to load a reference to `output`.

nhz2 commented on base/stat.jl:
```diff
@@ -301,7 +301,7 @@ function filemode_string(mode)
         complete && write(str, "-")
     end
-    return String(take!(str))
+    return unsafe_takestring!(str)
```
Why is this guaranteed to be safe?

jakobnissen replied:
Similar here, we can guarantee no reference to `str` is leaked out of the function.

nhz2 commented on base/strings/string.jl:
```diff
@@ -83,6 +78,30 @@ function String(v::Vector{UInt8})
+"Create a string re-using the memory, if possible.
+Mutating the memory after calling this function is undefined behaviour."
+function unsafe_takestring(m::Memory{UInt8})
+    ccall(:jl_genericmemory_to_string, Ref{String}, (Any, Int), m, length(m))
```
Why is this missing the:

```julia
len = length(m)
len == 0 && return ""
```

jakobnissen replied:
I can't remember, possibly an oversight. Added.

nhz2 commented on stdlib/Base64/src/Base64.jl:
```diff
@@ -8,7 +8,7 @@ a method to represent binary data using text, common on the web.
 module Base64
 
-using Base: require_one_based_indexing
+using Base: require_one_based_indexing, unsafe_takestring!
```
Why is this being imported here?

jakobnissen replied:
Removed.

nhz2 commented on stdlib/Markdown/src/Markdown.jl:
```diff
@@ -9,7 +9,7 @@ literals `md"..."` and `doc"..."`.
 module Markdown
 
-import Base: AnnotatedString, AnnotatedIOBuffer, show, ==, with_output_color, mapany
+import Base: AnnotatedString, AnnotatedIOBuffer, show, ==, with_output_color, mapany, unsafe_takestring!
```
Why is this being imported here?

jakobnissen replied:
Unnecessary, perhaps a leftover from an earlier change. Changed.

nhz2 commented on test/iobuffer.jl:
```diff
@@ -54,6 +54,26 @@ bufcontents(io::Base.GenericIOBuffer) = unsafe_string(pointer(io.data), io.size)
     @test_throws ArgumentError seek(io, 0)
 end
 
+@testset "takestring!" begin
```
This needs to be tested for all the different kinds of `IOBuffer`

jakobnissen replied:
Done. There are only two kinds of `IOBuffer`: The normal, seekable one (which may be writable or not), and the `PipeBuffer`.

jakobnissen commented on base/io.jl:
```diff
@@ -543,8 +543,8 @@ julia> rm("my_file.txt")
 readuntil(filename::AbstractString, delim; kw...) = open(io->readuntil(io, delim; kw...), convert(String, filename)::String)
-readuntil(stream::IO, delim::UInt8; kw...) = _unsafe_take!(copyuntil(IOBuffer(sizehint=16), stream, delim; kw...))
-readuntil(stream::IO, delim::Union{AbstractChar, AbstractString}; kw...) = String(_unsafe_take!(copyuntil(IOBuffer(sizehint=16), stream, delim; kw...)))
+readuntil(stream::IO, delim::UInt8; kw...) = takestring!(copyuntil(IOBuffer(sizehint=16), stream, delim; kw...))
```
Done

stevengj commented on base/iobuffer.jl:
```diff
@@ -489,6 +489,84 @@ function take!(io::IOBuffer)
+end
+
+# Fallback methods
+unsafe_takestring!(io::GenericIOBuffer) = takestring!(io)
```
Why not?

nhz2 replied:
I removed it and the tests passed locally. There are also no functions that call `unsafe_takestring!` on a `GenericIOBuffer`.

stevengj replied:
A problem could arise if you construct a `GenericIOBuffer` with a backing buffer that is not `Vector{UInt8}`.  I'm not sure if this is exercised by the tests.

nhz2 replied:
`unsafe_takestring!` is an internal function, so it shouldn't be an issue.

jakobnissen replied:
Done

stevengj commented on base/show.jl:
```diff
@@ -2718,7 +2718,7 @@ function type_depth_limit(str::String, n::Int; maxdepth = nothing)
         prev = di
     end
-    return String(take!(output))
+    return unsafe_takestring!(output)
```
Isn't it clearly safe because the `output` object does not escape?

nhz2 replied:
That isn't clear to me. Are `write(output, c)` and `write(output, c.del)` guaranteed not to escape `output`? Testing that seems out of scope for this PR.

stevengj replied:
Yes, because Base owns the methods `write(::IOBuffer, ::Char)` and `write(::IOBuffer, ::SubString{String})`.

(The same requirement is made in many other uses of `unsafe_takestring!` and `unsafe_take!` in Base … they are used whenever a function writes to an `IOBuffer` that is allocated internally to the function.  It seems strange to quibble about them here.)

stevengj replied:
(There is a legitimate worry about functions like `sprint` that were already using `_unsafe_take!`, but were passing it to user-defined methods that might conceivably save the buffer.)

nhz2 replied:
Okay, that makes sense.

stevengj commented on base/stat.jl:
```diff
@@ -301,7 +301,7 @@ function filemode_string(mode)
         complete && write(str, "-")
     end
-    return String(take!(str))
+    return unsafe_takestring!(str)
```
Same as above: this is safe because we are only calling methods we own, `write(::IOBuffer, ::Char)` and `write(::IOBuffer, ::String)` and don't return `str`.

vtjnash commented:
If it matters, this SGTM. Even if we keep around the mutating behavior of String(Vector{UInt8}) for now until v2, it still seems good for readability to add these more precise function names now.

nhz2 commented:
Bump @jakobnissen I think my suggestions fix all the currently failing tests.

jakobnissen commented on base/iobuffer.jl:
```diff
@@ -489,6 +489,84 @@ function take!(io::IOBuffer)
+        io.offset = 0
+    end
+    s
+end
```
I changed the function in the latest commit to handle the edge cases. I don't think there are many - just if the buffer is writable or not (especially after my IOBuffer refactoring)

jakobnissen commented:
Thanks for the thorough review, @nhz2. I addressed your points and rebased on master. As soon as all tests pass, this PR is good to go.

nhz2 commented on base/iobuffer.jl:
````diff
@@ -783,6 +783,80 @@ function take!(io::IOBuffer)
+true
+```
+
+!!! compat "Julia 1.12"
+    This function requires at least Julia 1.12.
````
```suggestion
!!! compat "Julia 1.13"
    This function requires at least Julia 1.13.
```

nhz2 reviewed (changes requested):
Apart from the compat version number, this looks good.

jakobnissen commented:
CI failures are unrelated.

stevengj commented on base/io.jl:
```diff
@@ -616,8 +616,7 @@ Logan
     open(io -> readline(io; keep), filename)
-readline(s::IO=stdin; keep::Bool=false) =
-    String(_unsafe_take!(copyline(IOBuffer(sizehint=16), s; keep)))
+readline(s::IO=stdin; keep::Bool=false) = takestring!(copyline(IOBuffer(sizehint=16), s; keep))
```
```suggestion
readline(s::IO=stdin; keep::Bool=false) = unsafe_takestring!(copyline(IOBuffer(sizehint=16), s; keep))
```

jakobnissen replied:
There is some discussion to be had, since `IO` is generic here, an implementation of `copyline` could store a reference to the `IOBuffer` somewhere and use it later. However, that does seem unlikely.

stevengj replied:
Ah, okay.  We were using `_unsafe_take!` before so I wasn't sure why it was removed.

stevengj commented on base/iostream.jl:
```diff
@@ -469,7 +469,7 @@ function readuntil_string(s::IOStream, delim::UInt8, keep::Bool)
 readuntil(s::IOStream, delim::AbstractChar; keep::Bool=false) =
     isascii(delim) ? readuntil_string(s, delim % UInt8, keep) :
-    String(_unsafe_take!(copyuntil(IOBuffer(sizehint=70), s, delim; keep)))
+    takestring!(copyuntil(IOBuffer(sizehint=70), s, delim; keep))
```
```suggestion
    unsafe_takestring!(copyuntil(IOBuffer(sizehint=70), s, delim; keep))
```

stevengj commented on base/strings/basic.jl:
```diff
@@ -678,7 +678,7 @@ function filter(f, s::AbstractString)
         f(c) && write(out, c)
     end
-    String(_unsafe_take!(out))
+    takestring!(out)
```
```suggestion
    unsafe_takestring!(out)
```

stevengj commented on base/strings/io.jl:
```diff
@@ -112,7 +112,7 @@ function sprint(f::Function, args...; context=nothing, sizehint::Integer=0)
         f(s, args...)
     end
-    String(_unsafe_take!(s))
+    takestring!(s)
```
```suggestion
    unsafe_takestring!(s)
```

stevengj commented on base/strings/io.jl:
```diff
@@ -146,7 +146,7 @@ function print_to_string(xs...)
         print(s, x)
     end
-    String(_unsafe_take!(s))
+    takestring!(s)
```
```suggestion
    unsafe_takestring!(s)
```

stevengj commented on base/strings/io.jl:
```diff
@@ -164,7 +164,7 @@ function string_with_env(env, xs...)
         print(env_io, x)
     end
-    String(_unsafe_take!(s))
+    takestring!(s)
```
```suggestion
    unsafe_takestring!(s)
```

DilumAluthge commented:
Looks like this is going to go through another round of triage discussion.

gbaraldi commented:
Triage likes this.

oscardssmith commented:
@jakobnissen thanks so much for all your work here! The one request from triage was a rebase to 1 commit that implements the new functions and a 2nd that switches uses/docs over.

jakobnissen committed:
Implement takestring!

This function creates a string from an IOBuffer or a Vector, emptying the input
object and reusing the memory where possible.

jakobnissen committed:
Use takestring! in Base and docs

This commit switches the usage of the pattern `String(take!(::IOBuffer))` to use
the new `takestring!` or `unsafe_takestring!` functions across Base, and in doc-
strings.
Not all occurrences in e.g. tests are switched over, as this would consistute a
lot of code churn for no real purpose.

JeffBezanson commented on base/iobuffer.jl:
```diff
@@ -783,6 +783,80 @@ function take!(io::IOBuffer)
+        io.size = 0
+        io.offset_or_compacted = 0
+        s
+    else
```
Why is `jl_genericmemory_to_string` (the no-copy path) only reachable if the io is writable?

nhz2 replied:
Currently, `takestring!` is mainly used for `IOBuffer`s that are used as string builders, so they are already writable.
Also, the last time I checked, calling `jl_genericmemory_to_string` can make the reading from the memory invalid: https://github.com/JuliaLang/julia/pull/54372#issuecomment-2248148052 , though this might have changed since then.

jakobnissen replied:
When IOBuffer is not writable, the IOBuffer doesn't take ownership of the memory, and allows it to be shared, e.g. between multiple IOBuffers (and that also makes IOBuffer more safe in general since the user can keep a reference to the memory they wrapped the IOBuffer around). Therefore, we need to copy it out.
Luckily, as Nathan said, that's usually not a performance problem for takestring! where the iobuffer is mutable.

IanButterworth replied:
@JeffBezanson it wasn't clear whether this is blocking but given triage approval, I went ahead

IanButterworth commented:
Assuming `don't squash` from the triage request

DilumAluthge commented:
Removing `merge me` label because there still seems to be a review conversation ongoing: https://github.com/JuliaLang/julia/pull/54372#discussion_r2064537928

nsajko commented:
There are conflicts.

nsajko commented:
Should `unsafe_takestring` be renamed to `unsafe_takestring!`?

jakobnissen commented:
> Should unsafe_takestring be renamed to unsafe_takestring!?

Sure, but let's do this in another PR.

jakobnissen commented:
Bump
